addons/web/static/tests/utils.js

```diff
@@ -644,3 +644,66 @@ class Contains {
 export async function contains(selector, options) {
     await new Contains(selector, options).run();
 }
+
+const stepState = {
+    expectedSteps: null,
+    deferred: null,
+    timeout: null,
+    currentSteps: [],
+
+    clear() {
+        clearTimeout(this.timeout);
+        this.timeout = null;
+        this.deferred = null;
+        this.currentSteps = [];
+        this.expectedSteps = null;
+    },
+
+    check({ crashOnFail = false } = {}) {
+        const success =
+            this.expectedSteps.length === this.currentSteps.length &&
+            this.expectedSteps.every((s, i) => s === this.currentSteps[i]);
+        if (!success && !crashOnFail) {
+            return;
+        }
+        QUnit.config.current.assert.verifySteps(this.expectedSteps);
+        if (success) {
+            this.deferred.resolve();
+        } else {
+            this.deferred.reject(new Error("Steps do not match"));
+        }
+        this.clear();
+    },
+};
+
+/**
+ * Indicate the completion of a test step. This step must then be verified by
+ * calling `assertSteps`.
+ *
+ * @param {string} step
+ */
+export function step(step) {
+    registerCleanup(() => stepState.clear());
```

Comment:
Maybe register it only once in `QUnit.testStart`?

Also I was suggesting to use `check` with `crashOnFail` inside cleanup rather than `clear` to also get crashes/error messages at the end of tests (at least if there are `expectedSteps` or `currentSteps`).

Reply:
Done, should be fine now! (only if they are `expectedSteps` since `currentSteps` is covered by `assert.step`)

Reply:
I'm not sure to understand why not `currentSteps` also.

If we just use a `step()` and then the test ends, we would want to also clear and crash.

Reply:
`check` assumes `expectedSteps` is set, there is no reason to call it otherwise. If someone called `step(...)` without `assertSteps` (the case you're mentionning), Qunit will fail anyway `expected verifySteps to be called...`. But yes the state should be cleared anyway.